deliveries/solidity-trie-verifier-milestone_1.md

```diff
@@ -0,0 +1,35 @@
+| 2 | KeySpacedDB | [https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L99](https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L99) | Optimized as [https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L106](https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L106) in `ReadChildProofCheck` |
+| 3 | MemoryDB | [https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L37](https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/MerklePatricia.sol#L37), [https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/trie/TrieDB.sol#L8](https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/trie/TrieDB.sol#L8)| We have replaced the MemoryDB which would have been a hashmap with an in-memory array because it is much cheaper to lookup items in an in-memory array as compared to using a mapping. Solidity does not support in-memory mappings, and they can only be created in storage. With this change, we have successfully reduced the storage costs and improved the overall efficiency. |
+| 4 | NodeCodec | [https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/substrate/SubstrateTrieDB.sol](https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/substrate/SubstrateTrieDB.sol) | |
+| 5 | Node | [https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/Node.sol#L9](https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/Node.sol#L9) | |
+| 6 | NodePlan | [https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/trie/Node.sol#L26](https://github.com/polytope-labs/solidity-merkle-trees/blob/82698e828b883eeb1ee0f658956c0c0ad26f5f49/src/trie/Node.sol#L26) | |
+| 7 | NodeHeader | [https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/Node.sol](https://github.com/polytope-labs/solidity-merkle-trees/blob/main/src/trie/Node.sol) | |
```

Comment:
Could you add more comments to this file? In particular, it's not clear to me where the[ `NodeHeader`](https://github.com/paritytech/substrate/blob/129fee774a6d185d117a57fd1e81b3d0d05ad747/primitives/trie/src/node_header.rs#L26) is implemented here (see specs in the proposal).

Reply:
@seunlanlege ok thx. Could you add some more comments though? I think it'd be very useful for anybody who wants to use your library.

Reply:
@Doordashcon 

> the 7. | NodeHeader has been updated

Are you sure you pushed the changes? I don't see any commits that are newer than 1 week.

> would also like to highlight that the [EthereumTrieDB.sol](https://github.com/polytope-labs/solidity-merkle-trees/blob/9ec345c82b9adbe49c024a29588e5ac631976639/src/trie/ethereum/EthereumTrieDB.sol) is part of ongoing improvements to the greater Solidity Merkle Trees library

thanks for the info, I'll ignore that file in that case for the scope of this review

Reply:
Hi @takahser
 > Are you sure you pushed the changes? I don't see any commits that are newer than 1 week.
 
Maybe some issues with git on my end, here is the commit [tag](https://github.com/w3f/Grant-Milestone-Delivery/pull/774/commits/d6130b18d8765ec46716f86065942e2c978fa98a) for better visibility.

Reply:
@Doordashcon, sorry for the misunderstanding, I'm talking about the https://github.com/polytope-labs/solidity-merkle-trees repo. I didn't see the update where you added more inline comments to the code. This concerns that part of the core logic that are not well documented using inline comments and the https://github.com/polytope-labs/solidity-merkle-trees/blob/9ec345c82b9adbe49c024a29588e5ac631976639/src/trie/substrate/SubstrateTrieDB.sol file in particular. See also some other files I mentioned in my [evaluation](https://github.com/w3f/Grant-Milestone-Delivery/blob/b7039469d2293676b5139baecb2710fa08f5ba97/evaluations/solidity-merkle-trees_1_takahser.md#general-notes).

Reply:
Hi @takahser, also highlighting inline documentation was not stipulated as part of the application/contract, this may be made as a `good first issue` for outside contributors or internally much later.

Reply:
@Doordashcon
> inline documentation was not stipulated as part of the application/contract

Fair point, I didn't notice this earlier, sorry. Usually this is part of the contract, but for some reason it's missing in yours. Hence, I'll give it a pass though I'd like to highlight that I still believe it'd be good to add them.